File: Data Structure and Algorithm/Python/kwarraylist.py
class KWArrayList: 
    name = ""
    theList = []
    size = 0
    capacity = 100

    def __init__(self, name):
        self.theList = [None] * self.capacity
        self.name = name

    def reallocate(self):
        self.capacity = self.capacity*2

    def add(self, item):
        if (self.size == self.capacity):
            self.reallocate()
        self.theList[self.size] = item
        self.size = self.size + 1

    def remove(self, item):
        theStat = False
        if (self.isEmpty()):
            return "Empty list! Cannot delete any item"
        for i in range(self.size):
            if (self.theList[i] == item):
                theStat = True
                for j in range(i, self.size - 1):
                    self.theList[j] = self.theList[j + 1]
                self.theList[self.size - 1] = None
                self.size = self.size - 1
                break
        if (theStat == False):
            return "Item not found in the list"
        else:
            return "Item removed successfully"
        
    def getSize(self):
        return self.size
    
    def get(self, index):
        if (index < 0 or index >= self.size):
            return "Index out of bounds"
        return self.theList[index]
    
    def search(self, item):
        for i in range(self.size):
            if (self.theList[i] == item):
                return i
        return -1
    
    def sort(self):
        for i in range(self.size):
            for j in range(0, self.size-i-1):
                if (self.theList[j] > self.theList[j+1]):
                    self.theList[j], self.theList[j+1] = self.theList[j+1], self.theList[j]

    def set(self, index, item):
        if (index < 0 or index >= self.size):
            return "Index out of bounds"
        self.theList[index] = item

    def replace(self, oldItem, newItem):
        index = self.search(oldItem)
        if (index == -1):
            return "Item not found in the list"
        self.theList[index] = newItem
        return "Item replaced successfully"
    
    def indexOf(self, item):
        return self.search(item)
    
    def printList(self):
        for i in range(self.size):
            print(self.theList[i], end=" ")
        print()

    def clear(self):
        self.theList = [None] * self.capacity
        self.size = 0

    def isEmpty(self):
        return self.size == 0
    
    def getName(self):
        return self.name

File: Data Structure and Algorithm/C#/KWArrayList.cs
using System;

public class KWArrayList<E> where E : class
{
	int size, capacity;
	E[] theList;

    public KWArrayList()
	{
		capacity = 100;
		theList = new E[capacity];
		size = 0;
	}

    public KWArrayList(int initialCapacity)
    {
        capacity = (initialCapacity < 1) ? 100 : initialCapacity;
        theList = new E[capacity];
        size = 0;
    }

    public int Size()
    {
        return size;
    }
    public bool IsFull()
    {
        return size == capacity-1;
    }
    public bool Range(int index)
    {
        return index >= 0 && index < size;
    }
    public E Get(int index)
    {
        if (!Range(index))
        {
            throw new IndexOutOfRangeException("Index out of range");
        }
        return theList[index];
    }
    public void reallocate()
    {
        capacity *= 2;
        E[] newList = new E[capacity];
        Array.Copy(theList, newList, size);
        theList = newList;
    }
    public void Add(E item)
    {
        if (IsFull())
        {
            reallocate();
        }
        theList[size] = item;
        size++;
    }
    public bool Contains(E item)
    {
        for (int i = 0; i < size; i++)
        {
            if (theList[i].Equals(item))
            {
                return true;
            }
        }
        return false;
    }
    public void Remove(E item)
    {
        if (IsEmpty())
        {
            throw new InvalidOperationException("List is empty");
        }
        for (int i = 0; i < size; i++)
        {
            if (theList[i].Equals(item))
            {
                for (int j = i; j < size - 1; j++)
                {
                    theList[j] = theList[j + 1];
                }
                theList[size - 1] = null;
                size--;
                return;
            }
        }
    }
    public void Clear()
    {
        for (int i = 0; i < size; i++)
        {
            theList[i] = null;
        }
        size = 0;
    }
    public void Print()
    {
        for (int i = 0; i < size; i++)
        {
            Console.WriteLine(theList[i]);
        }
    }
    public void Sort()
    {
        Array.Sort(theList, 0, size);
    }
    public void Reverse()
    {
        Array.Reverse(theList, 0, size);
    }
    public int IndexOf(E item)
    {
        for (int i = 0; i < size; i++)
        {
            if (theList[i].Equals(item))
            {
                return i;
            }
        }
        return -1;
    }
    public int LastIndexOf(E item)
    {
        for (int i = size - 1; i >= 0; i--)
        {
            if (theList[i].Equals(item))
            {
                return i;
            }
        }
        return -1;
    }
    public int Capacity()
    {
        return capacity;
    }
    public int Count()
    {
        return Size();
    }
    public bool IsEmpty()
    {
        return size == 0;
    }
    public bool Remove()
    {
        if (IsEmpty())
        {
            return false;
        }
        theList[size] = null;
        size--;
        return true;
    }
    public void TrimToSize()
    {
        if (size < capacity)
        {
            E[] newList = new E[size];
            Array.Copy(theList, newList, size);
            theList = newList;
            capacity = size;
        }
    }
    public void Set(int index, E item)
    {
        if (index < 0 || index >= size)
        {
            throw new IndexOutOfRangeException("Index out of range");
        }
        theList[index] = item;
    }
    public int Search(E item)
    {
        return IndexOf(item);
    }
    public bool RemoveAt(int index)
    {
        if (!Range(index))
        {
            throw new IndexOutOfRangeException("Index out of range");
        }
        for (int i = index; i < size - 1; i++)
        {
            theList[i] = theList[i + 1];
        }
        theList[size - 1] = null;
        size--;
        return true;
    }
    public bool RemoveAll(E item)
    {
        bool removed = false;
        for (int i = 0; i < size; i++)
        {
            if (theList[i].Equals(item))
            {
                RemoveAt(i);
                i--;
                removed = true;
            }
        }
        return removed;
    }
    public bool RetainAll(E[] items)
    {
        bool modified = false;
        for (int i = 0; i < size; i++)
        {
            if (Array.IndexOf(items, theList[i]) == -1)
            {
                RemoveAt(i);
                i--;
                modified = true;
            }
        }
        return modified;
    }
    public bool ContainsAll(E[] items)
    {
        foreach (E item in items)
        {
            if (!Contains(item))
            {
                return false;
            }
        }
        return true;
    }
    public bool AddAll(E[] items)
    {
        foreach (E item in items)
        {
            Add(item);
        }
        return true;
    }
    public bool Replace(E itemOld, E itemNew)
    {
        int index = IndexOf(itemOld);
        if (index != -1)
        {
            Set(index, itemNew);
            return true;
        }
        return false;
    }
}
